crates/escrow/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+    /// The pallet's configuration trait.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Convert the block number into a balance.
+        type BlockNumberToBalance: Convert<Self::BlockNumber, BalanceOf<Self>>;
+
+        /// The currency trait.
+        type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>
+            + ReservableCurrency<Self::AccountId>;
+
+        /// All future times are rounded by this.
+        #[pallet::constant]
+        type Span: Get<Self::BlockNumber>;
+
+        /// The maximum time for locks.
+        #[pallet::constant]
+        type MaxPeriod: Get<Self::BlockNumber>;
+
+        /// Weight information for the extrinsics in this module.
+        type WeightInfo: WeightInfo;
+    }
+
+    // The pallet's events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        Deposit {
+            who: T::AccountId,
+            amount: BalanceOf<T>,
+            unlock_height: T::BlockNumber,
+        },
+        Withdraw {
+            who: T::AccountId,
+            amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidAmount,
+        InvalidHeight,
+        LockNotFound,
+        LockNotExpired,
+        LockHasExpired,
+        InsufficientFunds,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+    #[pallet::storage]
+    #[pallet::getter(fn locked_balance)]
+    pub type Locked<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, LockedBalance<BalanceOf<T>, T::BlockNumber>, ValueQuery>;
+
+    #[pallet::storage]
+    pub type Epoch<T: Config> = StorageValue<_, T::Index, ValueQuery>;
+
+    #[pallet::storage]
+    pub type PointHistory<T: Config> =
+        StorageMap<_, Identity, T::Index, Point<BalanceOf<T>, T::BlockNumber>, ValueQuery>;
+
+    #[pallet::storage]
+    pub type UserPointHistory<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Identity,
+        T::Index,
+        Point<BalanceOf<T>, T::BlockNumber>,
+        ValueQuery,
+    >;
+
+    #[pallet::storage]
+    pub type UserPointEpoch<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, T::Index, ValueQuery>;
+
+    #[pallet::storage]
+    pub type SlopeChanges<T: Config> = StorageMap<_, Blake2_128Concat, T::BlockNumber, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(<T as Config>::WeightInfo::create_lock())]
+        #[transactional]
+        pub fn create_lock(
+            origin: OriginFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
```

Comment:
I might read this wrong, but wouldn't that allow to lock any currency? We only want to allow to lock KINT on kintsugi and INTR on interlay.

Reply:
No, `BalanceOf` is essentially a u128. In contrast to the `Amount` struct that we use throughout much of the code, it does not include the currencyid. The used currency is defined by the `Currency` type in the config